### PR TITLE
Improvements to libgit2

### PR DESCRIPTION
**Pull request by kshyatt**
I know there's a lot going on here.
- I documented some functions inline that were confusing to me, and perhaps to others.
- I added a few utility functions that were "missing" - constructors, a few API functions, etc.
- I fixed some things that were not `BoundsError`-ing that really need to be, cc @tkelman 
- I added a bunch of tests

### CODE REVIEW

File: base/libgit2/blob.jl [hunk 1]
```diff
@@ -1,4 +1,16 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
+function GitBlob(path::AbstractString)
```
Comment (tkelman) [inline base/libgit2/blob.jl]: put at least one blank line after the license header

File: base/libgit2/blob.jl [hunk 2]
```diff
@@ -1,4 +1,16 @@
...
+    blob_ptr_ptr = Ref{Ptr{Void}}(C_NULL)
+    apath = abspath(path)
+    @check ccall((:git_blob_create_fromdisk, :libgit2), Ptr{Void},
+                  (Ptr{Ptr{Void}}, Ptr{UInt8}), blob_ptr_ptr, apath)
+    return GitBlob(blob_ptr_ptr[])
```
Comment (tkelman) [inline base/libgit2/blob.jl]: should this check for null before dereferencing or trying to create a GitBlob?

File: base/libgit2/diff.jl [hunk 3]
```diff
@@ -36,9 +36,12 @@ function Base.count(diff::GitDiff)
 end
 
 function Base.getindex(diff::GitDiff, i::Csize_t)
+    if i < 1 || i >= count(diff)
```
Comment (tkelman) [inline base/libgit2/diff.jl]: is it a bounds error to ask for `diff[count(diff)]` or is that the last element?

File: base/libgit2/index.jl [hunk 4]
```diff
@@ -112,9 +112,12 @@ function Base.count(idx::GitIndex)
 end
 
 function Base.getindex(idx::GitIndex, i::Csize_t)
+    if i < 1 || i >= count(idx)
+        throw(BoundsError())
```
Comment (tkelman) [inline base/libgit2/index.jl]: Same question as above, and should pass the `BoundsError` the object and the index. Same for the error in `Base.getindex(status::GitStatus, i::Csize_t)`.

File: test/libgit2.jl [hunk 5]
```diff
@@ -130,6 +140,47 @@ mktempdir() do dir
...
+                @test isdir(cache_repo)
+                @test isdir(joinpath(cache_repo, ".git"))
+
+                # set a remote branch
+                branch = "upstream2"
+                LibGit2.GitRemote(repo, branch, repo_url,"+refs/heads/$branch:refs/remotes/origin/$branch") |> finalize
+
+                config = joinpath(cache_repo, ".git", "config")
+                lines = split(open(readstring, config, "r"), "\n")
+                @test any(map(x->x == "[remote \"upstream\"]", lines))
```
Comment (tkelman) [inline test/libgit2.jl]: don't need the `map`, since `any` can take a predicate

File: test/libgit2.jl [hunk 6]
```diff
@@ -130,6 +140,47 @@ mktempdir() do dir
...
+            try
+                @test isdir(cache_repo)
+                @test isdir(joinpath(cache_repo, ".git"))
+
+                # set a remote branch
+                branch = "upstream3"
+                LibGit2.GitRemoteAnon(repo, repo_url) |> finalize
+
+                config = joinpath(cache_repo, ".git", "config")
+                lines = split(open(readstring, config, "r"), "\n")
```
Comment (tkelman) [inline test/libgit2.jl]: I'd want to check that these don't fail if they get windows newlines in them (or check that it's impossible for that to be the case)

File: test/libgit2.jl [hunk 7]
```diff
@@ -229,6 +281,10 @@ mktempdir() do dir
                     @test auth.email == test_sig.email
                 end
 
+                @test LibGit2.iscommit(string(commit_oid1),repo)
+                @test LibGit2.iscommit(string(commit_oid2),repo)
+                @test LibGit2.is_ancestor_of(string(commit_oid1),string(commit_oid2),repo)
+                @test LibGit2.revcount(repo, string(commit_oid1), string(commit_oid2)) == (-1,1)
```
Comment (tkelman) [inline test/libgit2.jl]: the inconsistency in argument order between these is unfortunate

File: base/libgit2/diff.jl (hunk 3 above)
Comment (kshyatt) [inline base/libgit2/diff.jl]: We're actually going to ask for `count(diff) - 1`

Comment (tkelman) [inline base/libgit2/diff.jl]: but `i` in Julia is 1-based, the subtraction by 1 happens in the ccall

Comment (kshyatt) [inline base/libgit2/diff.jl]: Oh, I see, you mean it should be `>` rather than `>=`

Comment (tkelman) [inline base/libgit2/diff.jl]: right, if it's not an error to ask for that last element

File: test/libgit2.jl (hunk 5 above)
Comment (kshyatt) [inline test/libgit2.jl]: IS that true for the test I copied this from as well?

Comment (tkelman) [inline test/libgit2.jl]: if it was of the form `any(map(f, collection))` then yeah you can replace that with `any(f, collection)`

File: base/libgit2/rebase.jl [hunk 8]
```diff
@@ -21,9 +21,12 @@ function current(rb::GitRebase)
 end
 
 function Base.getindex(rb::GitRebase, i::Csize_t)
+    if i < 1 || i >= count(rb)
```
Comment (tkelman) [inline base/libgit2/rebase.jl]: what about here, is asking for the last element valid?

Comment (kshyatt) [inline base/libgit2/rebase.jl]: Knew I missed one!

**Comment by kshyatt** [comment]
AppVeyor seems to have timed out. @tkelman, is this ok to go?

**Comment by tkelman** [comment]
Not if `repo = LibGit2.init(cache_repo)` obeys global gitconfigs and could ever create a CRNL-terminated config file, for which this would fail.

File: base/libgit2/blob.jl (hunk 2 above)
Comment (wildart) [inline base/libgit2/blob.jl]: If error is returned, the exception will be raised.

File: base/libgit2/blob.jl [hunk 9]
```diff
@@ -1,5 +1,18 @@
...
+end
+
+function owner(blob::GitBlob)
+    repo_ptr = @check ccall((:git_blob_owner, :libgit2), Ptr{Void}, (Ptr{Void},), blob.ptr)
+    return GitRepo(repo_ptr)
```
Comment (wildart) [inline base/libgit2/blob.jl]: Wrong, do not use `@check`. Test result for C_NULL.

File: base/libgit2/diff.jl [hunk 10]
```diff
@@ -36,9 +36,12 @@ function Base.count(diff::GitDiff)
 end
 
 function Base.getindex(diff::GitDiff, i::Csize_t)
+    if i < 1 || i > count(diff)
+        throw(BoundsError(diff, i))
+    end
     delta_ptr = ccall((:git_diff_get_delta, :libgit2), Ptr{Void},
                        (Ptr{Void}, Csize_t), diff.ptr, i-1)
-    delta_ptr == C_NULL && return nothing
+    delta_ptr == C_NULL && throw(Error.GitError(Error.ERROR))
```
Comment (wildart) [inline base/libgit2/diff.jl]: I think, `BoundsError` is more appropriate here. C_NULL is returned if out of bounds ([ref](https://libgit2.github.com/libgit2/#HEAD/group/index/git_index_get_byindex)). Use the same exception in similar cases for `GitIndex`, etc.

File: base/libgit2/remote.jl [hunk 11]
```diff
@@ -88,3 +88,9 @@ function push{T<:AbstractString}(rmt::GitRemote, refspecs::Vector{T};
         !no_refs && finalize(sa)
     end
 end
+
+function name(rmt::GitRemote)
+    str_ptr = ccall((:git_remote_name, :libgit2), Cstring, (Ptr{Void}, ), rmt.ptr)
+    str_ptr == C_NULL && throw(Error.GitError(Error.ERROR))
```
Comment (wildart) [inline base/libgit2/remote.jl]: No need for exception, return `nothing`, because  C_NULL is returned for in-memory remotes.

**Comment by kshyatt** [comment]
@tkelman @wildart I've updated the PR - thoughts?

File: base/libgit2/remote.jl [hunk 12]
```diff
@@ -88,3 +88,9 @@ function push{T<:AbstractString}(rmt::GitRemote, refspecs::Vector{T};
         !no_refs && finalize(sa)
     end
 end
+
+function name(rmt::GitRemote)
+    str_ptr = ccall((:git_remote_name, :libgit2), Cstring, (Ptr{Void}, ), rmt.ptr)
+    str_ptr == C_NULL && return nothing
```
Comment (tkelman) [inline base/libgit2/remote.jl]: empty string or error?

Comment (kshyatt) [inline base/libgit2/remote.jl]: Also, `rmt` may be anonymous, in which case it has no `name`, right? This seems like a prime candidate for judicious application of `Nullable`...

Comment (wildart) [inline base/libgit2/remote.jl]: Or empty string

File: test/libgit2.jl (hunk 6 above)
Comment (kshyatt) [inline test/libgit2.jl]: So I should just do a `run(unix2dos(some crap))` or something here?

Comment (tkelman) [inline test/libgit2.jl]: that would modify the file - maybe do something with `eachline` and `chomp` ?

Comment (wildart) [inline test/libgit2.jl]: Is `.git/config` file on windows really has `\r` symbols?

Comment (tkelman) [inline test/libgit2.jl]: it might if you edit it with something dumb

**Comment by kshyatt** [comment]
Superseded by #19839 and friends I think.

**Pull request closed without merging**